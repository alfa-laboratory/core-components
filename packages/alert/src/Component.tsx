import React, { FC } from 'react';
import cn from 'classnames';

import { Plate, PlateProps } from '@alfalab/core-components-plate';

import styles from './index.module.css';

export type AlertProps = Omit<PlateProps, 'foldable' | 'defaultFolded' | 'leftAddons'>;

export const Alert: FC<AlertProps> = ({ className, title, ...restProps }) => (
    <Plate
        className={cn(styles.component, className)}
        buttonsClassName={styles.buttons}
        title={title ? <span className={styles.title}>{title}</span> : null}
        {...restProps}
    />
);
