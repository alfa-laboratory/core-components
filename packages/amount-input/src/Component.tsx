import cn from 'classnames';
import React, { useState, useEffect, useCallback, forwardRef } from 'react';
import { Input, InputProps } from '@alfalab/core-components-input';
import { FormControl } from '@alfalab/core-components-form-control';
import { THINSP, formatAmount, getCurrencySymbol } from '@alfalab/utils';
import { CurrencyCodes } from '@alfalab/data';
import { getFormatedValue, getAmountValueFromStr } from './utils';

import styles from './index.module.css';

export type AmountInputProps = Omit<InputProps, 'value' | 'onChange' | 'type'> & {
    /**
     * Денежное значение в минорных единицах
     * Значение null - значит не установлено
     */
    value?: number | null;

    /**
     * Валюта
     */
    currency?: CurrencyCodes;

    /**
     * Минорные единицы
     */
    minority?: number;

    /**
     * Жир
     */
    bold?: boolean;

    /**
     * Обработчик события изменения значения
     */
    onChange?: (
        e: React.ChangeEvent<HTMLInputElement>,
        payload: {
            /**
             * Денежное значение в минорных единицах
             * Значение null - значит не установлено
             */
            value: number | null;
            /**
             * Значение инпута
             */
            valueString: string;
        },
    ) => void;
};

/**
 * Компонент для ввода денежных значений
 */
export const AmountInput = forwardRef<HTMLInputElement, AmountInputProps>(
    (
        {
            value = null,
            minority = 100,
            currency = 'RUR',
            placeholder = `0\u2009${getCurrencySymbol(currency) || ''}`,
            bold = true,
            className,
            focusedClassName,
            dataTestId,
            clear = false,
            onChange,
            onBlur,
            onFocus,
            onClear,
            ...restProps
        },
        ref,
    ) => {
        const [focused, setFocused] = useState(false);
        const [inputValue, setInputValue] = useState<string>(
            formatAmount({
                // TODO: поддержать nullable в utils
                value: value as number,
                currency,
                minority,
            }).formated,
        );
        const filled = Boolean(inputValue || focused);

        const currencySymbol = getCurrencySymbol(currency);

        useEffect(() => {
            const currentAmountValue = getAmountValueFromStr(inputValue, minority);
            if (currentAmountValue !== value) {
                return setInputValue(
                    formatAmount({
                        // TODO: поддержать nullable в utils
                        value: value as number,
                        currency,
                        minority,
                    }).formated,
                );
            }

            return () => undefined;
            // eslint-disable-next-line react-hooks/exhaustive-deps
        }, [value, currency, minority]);

        const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
            const input = e.target;
            const enteredValue = input.value.replace(/\s/g, '').replace('.', ',');
            const isCorrectEnteredValue = /(^[0-9]{1,9}(,([0-9]+)?)?$|^\s*$)/.test(enteredValue);

            if (isCorrectEnteredValue) {
                const newFormatedValue = getFormatedValue(enteredValue, currency, minority);

                if (newFormatedValue === inputValue) {
                    const caret = input.selectionStart;
                    window.requestAnimationFrame(() => {
                        input.selectionStart = caret;
                        input.selectionEnd = caret;
                    });
                } else {
                    /**
                     * Поддержка положения каретки
                     * Поскольку при форматировании введенного значения могут появляться символы типа пробела
                     * или запятая - каретка прыгает в конец и ее необходимо ставить в правильное место
                     */

                    // Узнаем длину оригинального инпута с условием обрезания лишних символов

                    const [head, tail] = input.value.split(/\.|,/);
                    let notFormattedEnteredValueLength = head.length;
                    if (tail) {
                        notFormattedEnteredValueLength += 1; // запятая или точка
                        notFormattedEnteredValueLength += tail.slice(0, 2).length; // только 2 символа в минорной части
                    }

                    const diff = newFormatedValue.length - notFormattedEnteredValueLength;
                    const caret = (input.selectionStart as number) + diff;
                    window.requestAnimationFrame(() => {
                        input.selectionStart = caret;
                        input.selectionEnd = caret;
                    });
                }

                setInputValue(newFormatedValue);
                if (onChange) {
                    onChange(e, {
                        value: getAmountValueFromStr(newFormatedValue, minority),
                        valueString: newFormatedValue,
                    });
                }
            } else {
                // Не двигаем каретку когда вставляется невалидный символ
                const caret = (input.selectionStart as number) - 1;
                window.requestAnimationFrame(() => {
                    input.selectionStart = caret;
                    input.selectionEnd = caret;
                });
            }
        };

        const handleInputFocus = useCallback(
            (event: React.FocusEvent<HTMLInputElement>) => {
                setFocused(true);

                if (onFocus) {
                    onFocus(event);
                }
            },
            [onFocus],
        );

        const handleInputBlur = useCallback(
            (event: React.FocusEvent<HTMLInputElement>) => {
                setFocused(false);

                if (onBlur) {
                    onBlur(event);
                }
            },
            [onBlur],
        );

        const handleClear = useCallback(
            (event: React.MouseEvent<HTMLButtonElement>) => {
                setInputValue('');

                if (onClear) {
                    onClear(event);
                }
            },
            [onClear],
        );

        const [majorPart, minorPart] = inputValue.split(',');

        return (
            <div
                className={cn(styles.container, {
                    [styles.bold]: bold,
                    [styles.focused]: focused,
                    [styles.filled]: filled,
                })}
            >
                <FormControl
                    {...restProps}
                    className={cn(styles.fakeValueWithCurrencyContainer)}
                    focused={focused}
                    filled={filled}
                >
                    <div>
                        <span className={styles.majorPart}>{majorPart}</span>

                        <span className={styles.minorPartAndCurrency}>
                            {minorPart !== undefined && `,${minorPart}`}
                            {THINSP}
                            {currencySymbol}
                        </span>
                    </div>
                </FormControl>

                <Input
                    {...restProps}
                    clear={clear}
                    placeholder={placeholder}
                    value={inputValue}
                    className={cn(styles.component, className)}
                    focusedClassName={focusedClassName}
                    inputClassName={styles.input}
                    onChange={handleChange}
                    onFocus={handleInputFocus}
                    onBlur={handleInputBlur}
                    onClear={handleClear}
                    dataTestId={dataTestId}
                    ref={ref}
                />
            </div>
        );
    },
);
