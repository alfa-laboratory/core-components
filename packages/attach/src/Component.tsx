import React, {
    InputHTMLAttributes,
    ButtonHTMLAttributes,
    useState,
    useCallback,
    useRef,
} from 'react';
import cn from 'classnames';
import mergeRefs from 'react-merge-refs';
import { Button, ButtonProps } from '@alfalab/core-components-button';
import { ProgressBar } from '@alfalab/core-components-progress-bar';
import { KeyboardFocusable } from '@alfalab/core-components-keyboard-focusable';
import { AttachmentSBlackIcon, AttachmentMBlackIcon } from '@alfalab/icons-classic';
import { pluralize } from '@alfalab/utils';
import { truncateFilename } from './utils';

import styles from './index.module.css';

export type AttachProps = Omit<InputHTMLAttributes<HTMLInputElement>, 'size' | 'type' | 'value'> & {
    /**
     * Содержимое кнопки для выбора файла
     */
    buttonContent?: React.ReactNode;

    /**
     * Свойства для кнопки
     */
    buttonProps?: ButtonProps & ButtonHTMLAttributes<HTMLButtonElement>;

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Число символов, после которого имя файла будет обрезаться
     */
    maxFilenameLength?: number;

    /**
     * Текст для случая, когда файл не загружен
     */
    noFileText?: string;

    /**
     * Процент выполнения загрузки файла
     */
    progressBarPercent?: number;

    /**
     * Размер компонента
     */
    size?: 'xs' | 's' | 'm' | 'l';

    /**
     * Содержимое поля ввода, указанное по умолчанию. Принимает массив объектов типа File или null.
     */
    value?: File[] | null;

    /**
     * Идентификатор для систем автоматизированного тестирования
     */
    dataTestId?: string;
};

const MULTIPLE_TEXTS: [string, string, string] = ['файл', 'файла', 'файлов'];

export const Attach = React.forwardRef<HTMLInputElement, AttachProps>(
    (
        {
            size = 's',
            accept,
            buttonContent = 'Выберите файл',
            buttonProps = {},
            className,
            disabled,
            dataTestId,
            id,
            maxFilenameLength,
            multiple,
            noFileText = 'Нет файла',
            progressBarPercent,
            value,
            onChange,
            ...restProps
        },
        ref,
    ) => {
        const [files, setFiles] = useState(value || []);

        const inputRef = useRef<HTMLInputElement>(null);
        const labelRef = useRef<HTMLLabelElement>(null);
        const buttonRef = useRef<HTMLButtonElement>(null);

        const handleInputChange = useCallback(
            (event: React.ChangeEvent<HTMLInputElement>) => {
                if (event.target.files) {
                    setFiles(Array.from(event.target.files));
                }

                if (onChange) {
                    onChange(event);
                }
            },
            [onChange],
        );

        const handleButtonClick = useCallback(
            event => {
                if (labelRef.current) {
                    labelRef.current.click();
                }
                if (buttonRef.current) {
                    buttonRef.current.focus();
                }

                if (buttonProps.onClick) {
                    buttonProps.onClick(event);
                }
            },
            [buttonProps],
        );

        const handleClearClick = useCallback(() => {
            if (inputRef.current) {
                inputRef.current.value = '';
            }

            setFiles([]);
        }, []);

        const Icon = size === 'xs' ? AttachmentSBlackIcon : AttachmentMBlackIcon;

        const statusTextContent =
            files.length === 1 ? (
                truncateFilename(files[0].name, maxFilenameLength)
            ) : (
                <abbr title={files.map(file => file.name).join()}>
                    {files.length} {pluralize(files.length, ...MULTIPLE_TEXTS)}
                </abbr>
            );

        return (
            <div
                className={cn(
                    cn(
                        styles.component,
                        styles[size],
                        {
                            [styles.disabled]: disabled,
                        },
                        className,
                    ),
                )}
            >
                <Button
                    {...buttonProps}
                    size={size}
                    disabled={disabled}
                    view={(buttonProps && buttonProps.view) || 'outlined'}
                    leftAddons={
                        (buttonProps && buttonProps.leftAddons) || (
                            <Icon className={cn(styles.icon)} />
                        )
                    }
                    onClick={handleButtonClick}
                    ref={buttonRef}
                >
                    <span>{buttonContent}</span>
                </Button>
                <label className={cn(styles.label)} htmlFor={id} ref={labelRef}>
                    <input
                        {...restProps}
                        className={cn(styles.control)}
                        accept={accept}
                        disabled={disabled}
                        id={id}
                        multiple={multiple}
                        tabIndex={-1}
                        type='file'
                        onChange={handleInputChange}
                        ref={mergeRefs([ref, inputRef])}
                        data-test-id={dataTestId}
                    />
                </label>
                {files && files.length > 0 ? (
                    <div className={cn(styles.file)}>
                        <span>{statusTextContent}</span>
                        <KeyboardFocusable>
                            {(ref, focused) => (
                                <button
                                    aria-label='очистить'
                                    type='button'
                                    className={cn(styles.clear, {
                                        [styles.focused]: focused,
                                    })}
                                    onClick={handleClearClick}
                                    ref={ref}
                                />
                            )}
                        </KeyboardFocusable>
                        {progressBarPercent && (
                            <ProgressBar
                                className={cn(styles.progressBar)}
                                value={progressBarPercent}
                                view='positive'
                            />
                        )}
                    </div>
                ) : (
                    <div className={cn(styles.noFile)}>{noFileText}</div>
                )}
            </div>
        );
    },
);

/**
 * Для отображения в сторибуке
 */
Attach.defaultProps = {
    size: 's',
    buttonContent: 'Выберите файл',
    noFileText: 'Нет файла',
};
