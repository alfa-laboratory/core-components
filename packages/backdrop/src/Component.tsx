import React, { MouseEvent } from 'react';
import cn from 'classnames';
import { TransitionProps } from 'react-transition-group/Transition';
import { CSSTransition } from 'react-transition-group';

import { CSSTransitionClassNames } from 'react-transition-group/CSSTransition';
import styles from './index.module.css';

export type BackdropProps = Partial<TransitionProps> & {
    /**
     * Прозрачный бэкдроп
     */
    invisible?: boolean;

    /**
     * Управляет видимостью компонента
     */
    open: boolean;

    /**
     * Обработчик клика по бэкдропу
     */
    onClose?: (event: MouseEvent<HTMLElement>) => void;

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Классы анимации
     *
     * http://reactcommunity.org/react-transition-group/css-transition#CSSTransition-prop-classNames
     */
    transitionClassNames?: string | CSSTransitionClassNames;

    /**
     * Идентификатор для систем автоматизированного тестирования
     */
    dataTestId?: string;

    /**
     * Прозрачность бэкдропа
     */
    opacity?: number;
};

export const Backdrop: React.FC<BackdropProps> = ({
    className,
    open = false,
    invisible = false,
    timeout = 200,
    children,
    onClose,
    dataTestId,
    transitionClassNames = styles,
    opacity = 1,
    ...restProps
}) => {
    return (
        <CSSTransition
            timeout={timeout}
            unmountOnExit={true}
            classNames={transitionClassNames}
            in={open}
            appear={true}
            {...restProps}
        >
            <div
                aria-hidden={true}
                onClick={onClose}
                data-test-id={dataTestId}
                className={cn(styles.backdrop, className, {
                    [styles.invisible]: invisible,
                })}
                style={{ opacity }}
            >
                {children}
            </div>
        </CSSTransition>
    );
};
