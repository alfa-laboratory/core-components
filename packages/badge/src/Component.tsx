import React from 'react';
import cn from 'classnames';

import styles from './index.module.css';

export type BadgeProps = {
    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     *  Вид компонента
     */
    view: 'icon' | 'count';

    /**
     * Размер компонента (только для view=count)
     */
    size?: 's' | 'm' | 'l';

    /**
     *  Видимость обводки вокруг иконки (только для view=icon)
     */
    visibleIconOutline?: boolean;

    /**
     * Контент компонента
     */
    content?: React.ReactElement | number;

    /**
     * Цветовое оформление иконки
     */
    iconColor?:
        | 'positive'
        | 'attention'
        | 'link'
        | 'negative'
        | 'tertiary'
        | 'secondary'
        | 'primary';

    /**
     * Идентификатор для систем автоматизированного тестирования
     */
    dataTestId?: string;
};

export const Badge = ({
    className,
    size = 'm',
    view,
    visibleIconOutline = false,
    content,
    iconColor,
    dataTestId,
}: BadgeProps) => {
    const isCountView = view === 'count';

    const isHidden = isCountView && typeof content === 'number' && content <= 0;
    const componentContent = isCountView && content && content >= 100 ? '99+' : content;

    return (
        <div
            className={cn(
                styles.component,
                styles[size],
                styles[view],
                iconColor && styles[iconColor],
                {
                    [styles.isHidden]: isHidden,
                    [styles.dot]: !content,
                    [styles.outline]: visibleIconOutline,
                },
                className,
            )}
            data-test-id={dataTestId}
        >
            {componentContent}
        </div>
    );
};
