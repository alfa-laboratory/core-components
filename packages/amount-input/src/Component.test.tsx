/**
 * @jest-environment jsdom-sixteen
 */

import React, { useState } from 'react';
import { render, fireEvent, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { CurrencyCodes } from '@alfalab/data';
import { AmountInput } from './index';
import { AmountInputProps } from './Component';

describe('AmountInput', () => {
    const THINSP = String.fromCharCode(8201);

    function renderAmountInput(
        value: number | null,
        currency: CurrencyCodes | null = 'RUR',
        props: AmountInputProps = {},
    ) {
        // TODO: почему тесты в кор компонентах цепляются к data-test-id вместо label?
        const dataTestId = 'test-id';
        const { getByTestId } = render(
            <AmountInput
                value={value}
                currency={currency as CurrencyCodes}
                minority={100}
                dataTestId={dataTestId}
                {...props}
            />,
        );

        const input = getByTestId(dataTestId) as HTMLInputElement;

        return input;
    }

    describe('Snapshots tests', () => {
        it('should match snapshot', () => {
            expect(render(<AmountInput />)).toMatchSnapshot();
            expect(
                render(<AmountInput value={1234567} currency='USD' minority={100} />),
            ).toMatchSnapshot();
        });
    });

    it('should use default placeholder', () => {
        const input = renderAmountInput(null);
        expect(input.placeholder).toBe(`0${THINSP}₽`);
    });

    it('should correctly render default placeholder if currency is empty', () => {
        const input = renderAmountInput(null, null);
        expect(input.placeholder).toBe(`0${THINSP}`);
    });

    it('should use passed placeholder', () => {
        const dataTestId = 'test-id';
        const { getByTestId } = render(
            <AmountInput
                value={null}
                currency='RUR'
                minority={100}
                dataTestId={dataTestId}
                placeholder='Сумма'
            />,
        );

        const input = getByTestId(dataTestId) as HTMLInputElement;
        expect(input.placeholder).toBe('Сумма');
    });

    it('should render passed amount', () => {
        const input = renderAmountInput(1234567);
        expect(input.value).toBe(`12${THINSP}345,67`);
    });

    it('should render passed amount without zero minor part', () => {
        const input = renderAmountInput(1234500);
        expect(input.value).toBe(`12${THINSP}345`);
    });

    it('should render empty input if passed amount.value is null', () => {
        const input = renderAmountInput(null);
        expect(input.value).toBe('');
    });

    it('should render 0 in input if passed amount.value is 0', () => {
        const input = renderAmountInput(0);
        expect(input.value).toBe('0');
    });

    it('should render passed decimal amount', () => {
        const input = renderAmountInput(1234567);
        expect(input.value).toBe(`12${THINSP}345,67`);
    });

    it('should allow input correct amounts', () => {
        const input = renderAmountInput(0);

        fireEvent.change(input, { target: { value: '123456' } });
        expect(input.value).toBe(`123${THINSP}456`);

        fireEvent.change(input, { target: { value: '123,' } });
        expect(input.value).toBe('123,');

        fireEvent.change(input, { target: { value: '123.' } });
        expect(input.value).toBe('123,');

        fireEvent.change(input, { target: { value: '123,4' } });
        expect(input.value).toBe('123,4');

        fireEvent.change(input, { target: { value: '123.4' } });
        expect(input.value).toBe('123,4');

        fireEvent.change(input, { target: { value: '123,45' } });
        expect(input.value).toBe('123,45');

        fireEvent.change(input, { target: { value: '123.45' } });
        expect(input.value).toBe('123,45');

        fireEvent.change(input, { target: { value: '123456789' } });
        expect(input.value).toBe(`123${THINSP}456${THINSP}789`);

        fireEvent.change(input, { target: { value: '123456789,12' } });
        expect(input.value).toBe(`123${THINSP}456${THINSP}789,12`);

        fireEvent.change(input, { target: { value: '123456789.12' } });
        expect(input.value).toBe(`123${THINSP}456${THINSP}789,12`);
    });

    it('should prevent input of incorrect amounts', () => {
        const input = renderAmountInput(1234567);

        fireEvent.change(input, { target: { value: 'f' } });
        expect(input.value).toBe(`12${THINSP}345,67`);

        fireEvent.change(input, { target: { value: '!' } });
        expect(input.value).toBe(`12${THINSP}345,67`);
    });

    it('should allow enter only integer values when integersOnly is true', async () => {
        const input = renderAmountInput(12345, 'RUR', { integersOnly: true });

        expect(input.value).toBe('123,45');

        await userEvent.type(input, '1');
        expect(input.value).toBe('123');

        await userEvent.type(input, '.');
        expect(input.value).toBe('123');

        await userEvent.type(input, ',');
        expect(input.value).toBe('123');

        await userEvent.type(input, '.50');
        expect(input.value).toBe(`12${THINSP}350`);

        input.focus();
        input.setSelectionRange(0, 3);
        await userEvent.paste(input, '123.456');
        expect(input.value).toBe('123');
    });

    it('should avoid inserting leading zero before number, but allow inserting zero', async () => {
        const input = renderAmountInput(null);
        await userEvent.type(input, '0');
        expect(input.value).toBe('0');
        await userEvent.type(input, '1234');
        expect(input.value).toBe(`1${THINSP}234`);
        input.focus();
        input.setSelectionRange(0, 0);
        await userEvent.type(input, '0', { initialSelectionStart: 0, initialSelectionEnd: 0 });
        expect(input.value).toBe(`1${THINSP}234`);
        fireEvent.change(input, { target: { value: '' } });
        await userEvent.type(input, '0');
        expect(input.value).toBe('0');
    });

    it('should allow replace minor part without deleting', async () => {
        const input = renderAmountInput(1234567);

        input.focus();
        input.setSelectionRange(7, 7);

        await userEvent.type(input, '8', { initialSelectionStart: 7, initialSelectionEnd: 7 });
        expect(input.value).toBe(`12${THINSP}345,86`);
    });

    it('should allow to past value with spaces', async () => {
        const input = renderAmountInput(null);

        await userEvent.paste(input, '1 23');
        expect(input.value).toBe('123');
    });

    it('should delete symbols on delete button press event', async () => {
        const input = renderAmountInput(null);

        await userEvent.type(input, '123,45');
        input.focus();
        input.setSelectionRange(1, 1);
        await userEvent.type(input, '{del}', { initialSelectionStart: 1, initialSelectionEnd: 1 });
        expect(input.value).toBe('13,45');
    });

    it('should allow set carret in the middle and enter decimal divider symbol', async () => {
        const input = renderAmountInput(null);

        await userEvent.type(input, '123456');

        input.setSelectionRange(4, 4);

        await userEvent.type(input, ',');
        /**
         * TODO: проверить положение карeтки
         * expect(input.selectionStart).toBe(4);
         */
        expect(input.value).toBe('123,45');

        // userEvent.type коряво внутри работает с кареткой поэтому после каждого ввода восстанавливаем каретку
        input.setSelectionRange(4, 4);
        await userEvent.type(input, '{backspace}');

        input.setSelectionRange(4, 4);
        await userEvent.type(input, '.');
        /**
         * TODO: проверить положение карeтки
         * expect(input.selectionStart).toBe(4);
         */
        expect(input.value).toBe('123,45');
    });

    it('should not delete any symbol when caret set after space and backspace pressed', async () => {
        const input = renderAmountInput(null);

        await userEvent.type(input, '1234');
        expect(input.value).toBe(`1${THINSP}234`);
        input.setSelectionRange(2, 2);
        await userEvent.type(input, '{backspace}');

        expect(input.value).toBe(`1${THINSP}234`);
    });

    it('should render new amount from props (looped value)', async () => {
        const dataTestId = 'test-id';
        let setAmountManually: (value: number, currency: CurrencyCodes, minority: number) => void;
        const HOCWithAmountInState = () => {
            const [value, setValue] = useState<number | null>(200);
            const [currency, setCurrency] = useState<CurrencyCodes>('RUR');
            const [minority, setMinority] = useState(100);

            /* eslint-disable no-shadow */
            setAmountManually = (
                value: number | null,
                currency: CurrencyCodes,
                minority: number,
            ) => {
                setValue(value);
                setCurrency(currency);
                setMinority(minority);
            };
            /* eslint-enable no-shadow */

            return (
                <AmountInput
                    value={value}
                    currency={currency}
                    minority={minority}
                    dataTestId={dataTestId}
                    onChange={(e, payload) => setValue(payload.value)}
                />
            );
        };

        const { getByTestId } = render(<HOCWithAmountInState />);

        const input = getByTestId(dataTestId) as HTMLInputElement;

        await userEvent.type(input, '{backspace}{backspace}{backspace}{backspace}5678');

        expect(input.value).toBe(`5${THINSP}678`);

        act(() => {
            setAmountManually(34567, 'USD', 100);
        });

        expect(input.value).toBe('345,67');

        await userEvent.clear(input);

        expect(input.value).toBe('');

        await userEvent.type(input, '0,');

        expect(input.value).toBe('0,');

        act(() => {
            setAmountManually(1, 'USD', 100);
        });

        expect(input.value).toBe('0,01');
    });

    describe('should emit value in minority on change event', () => {
        const dataTestId = 'test-id';

        const testCases = [
            {
                minority: 100,
                userInput: '1',
                expectedValue: 100,
            },
            {
                minority: 100,
                userInput: '1,1',
                expectedValue: 110,
            },
            {
                minority: 1000,
                userInput: '2',
                expectedValue: 2000,
            },
            {
                minority: 1000,
                userInput: '2,2',
                expectedValue: 2200,
            },
            {
                minority: 100,
                userInput: '9,12',
                expectedValue: 912,
            },
        ];

        testCases.forEach(({ minority, userInput, expectedValue }) => {
            it(`should emit event with value=${expectedValue} when minority=${minority} and userInput=${userInput}`, async () => {
                const handleChangeMock = jest.fn();
                const { getByTestId } = render(
                    <AmountInput
                        minority={minority}
                        dataTestId={dataTestId}
                        onChange={handleChangeMock}
                    />,
                );
                const input = getByTestId(dataTestId) as HTMLInputElement;
                await userEvent.paste(input, userInput);
                expect(handleChangeMock).toBeCalledWith(expect.anything(), {
                    value: expectedValue,
                    valueString: userInput,
                });
            });
        });
    });

    /**
     * + тест на адекватность (снапшот)
     * + тест на дефолтные значения (нужно разобраться про label)
     * - проброс пропсов
     * + 100003 вставить ',' после 100
     * + 100003 вставить '.' после 100
     * максимум 12 символов в мажорной части и не более 2х в минорной
     * + 1234 каретка перед двойкой backspace
     * + Тест на то что если amount зацикливать то все ок
     * - Тест при вставке невалидного символа каретка не двигается
     */
});
