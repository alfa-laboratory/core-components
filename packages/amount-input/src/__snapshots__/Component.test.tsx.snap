// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmountInput Snapshots tests should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container bold"
      >
        <div
          class="component s"
        >
          <div
            class="formControl component hasSuffix inner"
          >
            <div
              class="inputWrapper"
            >
              <div
                class="input"
              >
                <input
                  class="input input"
                  placeholder="0 ₽"
                  type="text"
                  value=""
                />
                <div
                  class="suffixContainer suffixContainer"
                >
                  <span
                    class="spacer"
                  />
                  <div
                    class="suffix"
                  >
                    
                    <span
                      class="minorPartAndCurrency"
                    >
                       
                      ₽
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
          
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container bold"
    >
      <div
        class="component s"
      >
        <div
          class="formControl component hasSuffix inner"
        >
          <div
            class="inputWrapper"
          >
            <div
              class="input"
            >
              <input
                class="input input"
                placeholder="0 ₽"
                type="text"
                value=""
              />
              <div
                class="suffixContainer suffixContainer"
              >
                <span
                  class="spacer"
                />
                <div
                  class="suffix"
                >
                  
                  <span
                    class="minorPartAndCurrency"
                  >
                     
                    ₽
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`AmountInput Snapshots tests should match snapshot 2`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container bold"
      >
        <div
          class="component s"
        >
          <div
            class="formControl component hasSuffix inner"
          >
            <div
              class="inputWrapper"
            >
              <div
                class="input"
              >
                <input
                  class="input input"
                  placeholder="0 ₽"
                  type="text"
                  value=""
                />
                <div
                  class="suffixContainer suffixContainer"
                >
                  <span
                    class="spacer"
                  />
                  <div
                    class="suffix"
                  >
                    
                    <span
                      class="minorPartAndCurrency"
                    >
                       
                      ₽
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
          
        </div>
      </div>
    </div>
    <div>
      <div
        class="container bold filled"
      >
        <div
          class="component s"
        >
          <div
            class="formControl component suffixVisible hasSuffix inner filled"
          >
            <div
              class="inputWrapper"
            >
              <div
                class="input"
              >
                <input
                  class="input input"
                  placeholder="0 $"
                  type="text"
                  value="12 345,67"
                />
                <div
                  class="suffixContainer suffixContainer"
                >
                  <span
                    class="spacer"
                  >
                    12 345,67
                  </span>
                  <div
                    class="suffix"
                  >
                    12 345
                    <span
                      class="minorPartAndCurrency"
                    >
                      ,67
                       
                      $
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
          
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container bold filled"
    >
      <div
        class="component s"
      >
        <div
          class="formControl component suffixVisible hasSuffix inner filled"
        >
          <div
            class="inputWrapper"
          >
            <div
              class="input"
            >
              <input
                class="input input"
                placeholder="0 $"
                type="text"
                value="12 345,67"
              />
              <div
                class="suffixContainer suffixContainer"
              >
                <span
                  class="spacer"
                >
                  12 345,67
                </span>
                <div
                  class="suffix"
                >
                  12 345
                  <span
                    class="minorPartAndCurrency"
                  >
                    ,67
                     
                    $
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
